src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/DeterministicShuffleShardSyncLeaderDecider.java

```diff
@@ -0,0 +1,184 @@
+     */
+    public DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager, int numPeriodicShardSyncWorkers) {
+        this(config, leaseManager, Executors.newScheduledThreadPool(LEADER_DECIDER_THREAD_COUNT),
+                numPeriodicShardSyncWorkers);
+    }
+
+    DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager, ScheduledExecutorService leaderElectionThreadPool,
+            int numPeriodicShardSyncWorkers) {
+        this.config = config;
+        this.leaseManager = leaseManager;
+        this.leaderElectionThreadPool = leaderElectionThreadPool;
+        this.numPeriodicShardSyncWorkers = numPeriodicShardSyncWorkers;
+    }
+
+    /*
+     * Shuffles the leases deterministically and elects numPeriodicShardSyncWorkers number of workers
+     * as leaders (workers that will perform shard sync).
+     */
+    private void electLeaders() {
+        try {
+            LOG.debug("Started leader election: " + System.currentTimeMillis());
+            List<KinesisClientLease> leases = leaseManager.listLeases();
+            List<String> uniqueHosts = leases.stream().map(KinesisClientLease::getLeaseOwner)
+                    .filter(owner -> owner != null).distinct().sorted().collect(Collectors.toList());
+
+            Collections.shuffle(uniqueHosts, new Random(DETERMINISTIC_SHUFFLE_SEED));
+            int numShardSyncWorkers = Math.min(uniqueHosts.size(), numPeriodicShardSyncWorkers);
+            // In case value is currently being read, we wait for reading to complete before updating the variable.
+            // This is to prevent any ConcurrentModificationException exceptions.
+            readWriteLock.writeLock().lock();
+            leaders = new HashSet<>(uniqueHosts.subList(0, numShardSyncWorkers));
+            LOG.debug("Elected leaders: " + String.join(", ", leaders));
+            LOG.debug("Completed leader election: " + System.currentTimeMillis());
+        } catch (DependencyException | InvalidStateException | ProvisionedThroughputException e) {
+            LOG.error("Exception occurred while trying to fetch all leases for leader election", e);
+        } catch (Throwable t) {
+            LOG.error("Unknown exception during leader election.", t);
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    // Utility methods to ensure we acquire a readLock in case an election has completed
+    // and some thread is trying to update leaders variable.
+    private boolean leadersNullOrEmpty() {
+        try {
+            readWriteLock.readLock().lock();
+            return CollectionUtils.isNullOrEmpty(leaders);
+        } finally {
+            readWriteLock.readLock().unlock();
+        }
+    }
+
+    private boolean workerIdInLeaders(String workerId) {
```

Comment:
Could you rename this method, it is not clear what is it doing.

Reply:
Changed to leadersContainsWorkerId()